### PR TITLE
[CI-7842]: Implement add-on change for ti for Python

### PR DESCRIPTION
**Pull request by happyhebaby**

### CODE REVIEW

File: ti/instrumentation/python/pytest.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+func (m *pytestRunner) GetCmd(ctx context.Context, tests []ti.RunnableTest, userArgs, workspace,
+	agentConfigPath, agentInstallDir string, ignoreInstr, runAll bool) (string, error) {
+	// Run all the tests
+	m.log.Infoln("Pytest GetCmd starting")
+
+	if runAll {
+		if ignoreInstr {
+			return strings.TrimSpace(fmt.Sprintf("%s %s", pytestCmd, userArgs)), nil
+		}
+		return strings.TrimSpace(
```
Comment (rutvijmehta-harness) [inline ti/instrumentation/python/pytest.go:60]: Define `"."` as constant and use it here

File: ti/instrumentation/python/pytest.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+		set[w] = struct{}{}
+		ut = append(ut, t.Class)
+	}
+	testStr := strings.Join(ut, ",")
+
+	if ignoreInstr {
+		return strings.TrimSpace(fmt.Sprintf("%s %s %s", pytestCmd, testStr, userArgs)), nil
+	}
+
+	return fmt.Sprintf("python3 python_agent.py %s --test_harness %s --disable_static",
```
Comment (rutvijmehta-harness) [inline ti/instrumentation/python/pytest.go:85]: Define `"."` as constant and use it he

File: ti/instrumentation/python/pytest.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+			return strings.TrimSpace(fmt.Sprintf("%s %s", pytestCmd, userArgs)), nil
+		}
+		return strings.TrimSpace(
+			fmt.Sprintf("python3 python_agent.py %s --test_harness %s --disable_static",
+				".", pytestCmd)), nil
+	}
+	if len(tests) == 0 {
+		return fmt.Sprintf("echo \"Skipping test run, received no tests to execute\""), nil
+	}
+	// Use only unique <package, class> tuples
```
Comment (rutvijmehta-harness) [inline ti/instrumentation/python/pytest.go:67]:
```suggestion
	// Use only unique classes
```

File: ti/instrumentation/python/pytest.go (hunk 2 above)
Comment (rutvijmehta-harness) [reply ti/instrumentation/python/pytest.go:85]: Where is `testStr` when not ignoring instrumentation?

File: ti/instrumentation/python/unittest.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+	"fmt"
+	"strings"
+
+	"github.com/harness/lite-engine/internal/filesystem"
+	"github.com/harness/lite-engine/ti"
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	unittestCmd = "unittest"
```
Comment (rutvijmehta-harness) [inline ti/instrumentation/python/unittest.go:20]: Camel case everywhere

File: ti/instrumentation/python/unittest.go [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+)
+
+type unittestRunner struct {
+	fs  filesystem.FileSystem
+	log *logrus.Logger
+}
+
+func NewUnittestRunner(log *logrus.Logger, fs filesystem.FileSystem) *unittestRunner { //nolint:revive
+	log.Infoln("NewUnitestRunner starting")
```
Comment (rutvijmehta-harness) [inline ti/instrumentation/python/unittest.go:29]: Typo

File: ti/instrumentation/python/unittest.go [hunk 6]
```diff
@@ -0,0 +1,87 @@
+// Copyright 2021 Harness Inc. All rights reserved.
```
Comment (rutvijmehta-harness) [inline ti/instrumentation/python/unittest.go:1]: All the same comments as `pytest.go`

File: ti/instrumentation/python/pytest.go (hunk 1 above)
Comment (rutvijmehta-harness) [reply ti/instrumentation/python/pytest.go:60]: Can you also look if it's possible to build the `agentArg` before hand and just add it here? For example: https://github.com/harness/lite-engine/blob/6d53379909358cdf3c9179e8ab64af5560771990/ti/instrumentation/java/maven.go#L57

File: ti/instrumentation/python/unittest.go [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+		set[w] = struct{}{}
+		ut = append(ut, t.Class)
+	}
+	testStr := strings.Join(ut, ",")
+
+	if ignoreInstr {
+		return strings.TrimSpace(fmt.Sprintf("%s %s %s", unittestCmd, testStr, userArgs)), nil
+	}
+
+	return fmt.Sprintf("python3 python_agent.py %s --test_harness %s --disable_static",
```
Comment (jamie-harness) [inline ti/instrumentation/python/unittest.go:85]: This command should have a list of test as well. How does your agent run if a list of tests is passed in?

File: ti/instrumentation/python/helper.go [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+)
+
+var (
+	getFiles     = utils.GetFiles
+	getWorkspace = external.GetWrkspcPath
+)
+
+// GetPythonTests returns list of RunnableTests in the workspace with python extension.
+// In case of errors, return empty list
+func GetPythonTests(testGlobs []string) ([]ti.RunnableTest, error) {
```
[the remark this answers is not in the document]
Comment (happyhebaby) [reply ti/instrumentation/python/helper.go:23]: Changed, I couldn't test e2e to verify the original implementation.

[the remark this answers is not in the document]
Comment (jamie-harness) [reply ti/instrumentation/python/helper.go:23]: As I mentioned in the Slack thread. Can you use the same logic as other languages and just put 
similar to this one
https://github.com/harness/harness-core/pull/47250/files#diff-428d42cf99a5feb27f8976c1e44817c8eec25c1614ea38517d16488cb4cc5229R31

**Review by jamie-harness — CHANGES_REQUESTED** [review]

**Review by jamie-harness — APPROVED** [review]

File: ti/instrumentation/python/unittest.go [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+func (m *unittestRunner) AutoDetectTests(ctx context.Context, workspace string, testGlobs []string) ([]ti.RunnableTest, error) {
+	pythonTests := GetPythonTests(workspace, testGlobs)
+	return pythonTests, nil
+}
+
+func (m *unittestRunner) GetCmd(ctx context.Context, tests []ti.RunnableTest, userArgs, workspace,
+	agentConfigPath, agentInstallDir string, ignoreInstr, runAll bool) (string, error) {
+	// Run all the tests
+	m.log.Infoln("Unittest GetCmd starting")
```
Comment (rutvijmehta-harness) [inline ti/instrumentation/python/unittest.go:50]: Log line doesn't add any benefit here. You can remove it.

File: ti/instrumentation/python/unittest.go [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+	}
+
+	testCmd := ""
+	if runAll {
+		if ignoreInstr {
+			return strings.TrimSpace(fmt.Sprintf("%s %s", unitTestCmd, userArgs)), nil
+		}
+		testCmd = strings.TrimSpace(fmt.Sprintf("python3 %s %s --test_harness %s",
+			scriptPath, currentDir, testHarness))
+		m.log.Infoln(fmt.Sprintf("testCmd: %s", testCmd))
```
Comment (rutvijmehta-harness) [inline ti/instrumentation/python/unittest.go:64]: Log line doesn't add any benefit here. You can remove it.

File: ti/instrumentation/python/unittest.go [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+		return testCmd, nil
+	}
+
+	if len(tests) == 0 {
+		return "echo \"Skipping test run, received no tests to execute\"", nil
+	}
+
+	// Use only unique <package, class> tuples
```
Comment (rutvijmehta-harness) [inline ti/instrumentation/python/unittest.go:72]: `// Use only unique classes`

File: ti/instrumentation/python/unittest.go [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+	}
+	testStr := strings.Join(ut, ",")
+
+	if ignoreInstr {
+		return strings.TrimSpace(fmt.Sprintf("%s %s %s", unitTestCmd, testStr, userArgs)), nil
+	}
+
+	testCmd = fmt.Sprintf("python3 %s %s --test_harness %s --test_files %s",
+		scriptPath, currentDir, testHarness, testStr)
+	m.log.Infoln(fmt.Sprintf("testCmd: %s", testCmd))
```
Comment (rutvijmehta-harness) [inline ti/instrumentation/python/unittest.go:92]: Log line doesn't add any benefit here. You can remove it.

File: ti/instrumentation/utils.go [hunk 13]
```diff
@@ -219,18 +219,18 @@ func downloadFile(ctx context.Context, path, url string, fs filesystem.FileSyste
 	// Create the file
 	out, err := fs.Create(path)
 	if err != nil {
-		return err
+		return fmt.Errorf("could not create directory: %s", err)
```
Comment (rutvijmehta-harness) [inline ti/instrumentation/utils.go:222]: The errors are handled upstream. Please revert this.

File: ti/instrumentation/utils.go [hunk 14]
```diff
@@ -219,18 +219,18 @@ func downloadFile(ctx context.Context, path, url string, fs filesystem.FileSyste
...
 	}
 	defer out.Close()
 
 	// Get the data
 	req, err := http.NewRequestWithContext(ctx, "GET", url, http.NoBody)
 	if err != nil {
-		return err
+		return fmt.Errorf("could not create request with context: %s", err)
```
Comment (rutvijmehta-harness) [inline ti/instrumentation/utils.go:229]: The errors are handled upstream. Please revert this.

File: ti/instrumentation/utils.go [hunk 15]
```diff
@@ -219,18 +219,18 @@ func downloadFile(ctx context.Context, path, url string, fs filesystem.FileSyste
...
 	}
 	resp, err := http.DefaultClient.Do(req)
 	if err != nil {
-		return err
+		return fmt.Errorf("could not make a request: %s", err)
```
Comment (rutvijmehta-harness) [inline ti/instrumentation/utils.go:233]: The errors are handled upstream. Please revert this.

File: ti/instrumentation/utils.go [hunk 16]
```diff
@@ -242,7 +242,7 @@ func downloadFile(ctx context.Context, path, url string, fs filesystem.FileSyste
 	// Writer the body to file
 	_, err = io.Copy(out, resp.Body)
 	if err != nil {
-		return err
+		return fmt.Errorf("could not copy: %s", err)
```
Comment (rutvijmehta-harness) [inline ti/instrumentation/utils.go:245]: The errors are handled upstream. Please revert this.

**Pull request merged**